oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/query/async/ElasticResultRowAsyncIterator.java

```diff
@@ -333,16 +339,24 @@ public void onSuccess(SearchResponse<ObjectNode> searchResponse) {
                 }
 
                 LOG.trace("Emitting {} search hits, for a total of {} scanned results", searchHits.size(), scannedRows);
+
+                Set<SearchHitListener> listenersWithHits = new HashSet<>();
+
                 for (Hit<ObjectNode> hit : searchHits) {
                     for (SearchHitListener l : searchHitListeners) {
-                        l.on(hit);
+                        if (l.on(hit)) {
+                            listenersWithHits.add(l);
+                        }
                     }
                 }
+                // if any listener has not processed any hit, it means we need to load more data since there could be
+                // listeners waiting for some results before triggering a new scan
+                boolean areAllListenersProcessed = listenersWithHits.containsAll(searchHitListeners);
```

Comment:
Nitpick: comparing the size of the sets is more efficient and produces the same result: if the sets have the same size, then all listeners got a hit.